src/ResourceTypeRegistry.ts

```diff
@@ -189,15 +203,10 @@ export default class ResourceTypeRegistry {
   }
 
   dbAdapter(typeName: string) {
-    const adapter = this.doGet("dbAdapter", typeName);
-    if(typeof adapter === 'undefined') {
-      throw new Error(
-        "Tried to get db adapter for a type registered without one. " +
-        "Every type must be registered with an adapter!"
-      );
-    }
-
-    return adapter;
+    // We can safely use the non null assertion because, if adapter is missing,
+    // attempting to call setRegistryDerivedOptions will fail in constructor.
+    // tslint:disable-next-line no-non-null-assertion
+    return this.doGet("dbAdapter", typeName)!;
```

Comment:
Is this comment accurate? Wouldn't `registry.dbAdapter('not a type')` give `undefined`?

Now that we're not calling `setRegistryDerivedOptions`, what should this function look like?

Reply:
> Is this comment accurate? Wouldn't `registry.dbAdapter('not a type')` give undefined?

Shit, you're totally right. And, by that logic, the non-null assertion in the `pagination` getter is incorrect too.

Your comment got me thinking about whether it's worth differentiating, in all the `ResourceTypeRegistry` getters, between the "`typeName` does not exist case" and the "`typeName`'s description doesn't define the given property case". I.e., right now

 ```ts
registry.beforeSave("made up type")
``` 

and 

```ts
registry.beforeSave("validTypeWithNoBeforeSave")
```

both return `undefined`.

Maybe that's fine, and we just make all the type annotations accurate, so that users know they could get `undefined` back, and then, when handling that, they can call `hasType` to figure out which case they're in. Any thoughts on that?

> Now that we're not calling `setRegistryDerivedOptions`, what should this function look like?

If we go with returning `undefined` in both the cases described above, I think the function body would just be:  `return this.doGet("dbAdapter", typeName);`. Then, to still notify users early about a missing adapter (which I think is nice), we'd throw in the constructor.

Reply:
> Your comment got me thinking about whether it's worth differentiating, in all the ResourceTypeRegistry getters, between the "typeName does not exist case" and the "typeName's description doesn't define the given property case".

I think the root cause here is that using `ResourceTypeRegistry.prototype.doGet()` prevents TypeScript from doing its thing.

```ts
registry.dbAdapter('schools')
registry.type('schools').dbAdapter
```

Although the second option is (slightly) more verbose, I think the verbosity is worthwhile here. Both steps (getting the type, getting the adapter) are steps where things could go wrong (be undefined). With this API, the following becomes possible:

```ts
interface ResourceTypeDef {
  dbAdapter: AdapterInstance,  // required
  parentTypeName?: string      // optional
  // ...
}

for (const name of registry.typeNames()) {
  registry.type(name)!.dbAdapter      // : AdapterInstance
  registry.type(name)!.parentTypeName // : string | undefined
}
```

I'm fast approaching the limit of my TypeScript knowledge here, but it may be possible to avoid the `!` for type lookups entirely with something like this:

```ts
type ValidTypeName = string;

class ResourceTypeRegistry {
  // ...
  typeNames(): ValidTypeName {}
  type(name: ValidTypeName): ResourceTypeDef {}
  // ...
}
```

And finally, if TypeScript supports it somehow, you could do some sort of method overloading to encode the fact that a `ValidTypeName` cannot result in `void` but a `string` can.

```ts
type(name: ValidTypeName): ResourceTypeDef
type(name: string): ResourceTypeDef | void
```

Reply:
> `registry.type('schools').dbAdapter`

You're right that that API enables clearer typings, so I'm glad we have it. Let's use that throughout the code when we know we have a valid type and we need to know that the adapter isn't undefined.

Let's also keep the other getters though (`beforeSave`, `beforeRender`, etc.) for convenience and to not have a bunch of breaking changes this close to v3. But let's just fix their typing by removing the incorrect non-null assertion operators.